File: receiver.go
package JTP

import (
	"encoding/binary"
	"net"
)

type Receiver struct {
	pConn    *net.UDPConn
	pSession *session
}

func NewReceiver() *Receiver {
	return &Receiver{}
}
func (r Receiver) log(str string, args ...interface{}) {
	Log("Receiver:"+str+"\n", args...)
}

func (r *Receiver) Dial(str_raddr string) error {
	raddr, err := net.ResolveUDPAddr("udp", str_raddr)
	if CheckErr(err) {
		return err
	}
	udpConn, err := net.DialUDP("udp", nil, raddr)
	if CheckErr(err) {
		return err
	}
	r.pConn = udpConn
	r.pSession = NewSession(*raddr)
	go r.recv()
	data := r.getReplyData(0, 0)
	_, err = r.pConn.Write(data)
	if CheckErr(err) {
		return err
	}
	return nil
}

func (r *Receiver) recv() {
	print("svr linten start")
	defer func() { Log("recv exit") }()
	for {
		data := make([]byte, 1024)
		l, addr, err := r.pConn.ReadFromUDP(data)
		if CheckErr(err) {
			continue
		}
		//fixme:模拟掉包
		debug := debugMissing[debugIdx%len(debugMissing)]
		debugIdx++
		if debug > 0 {
			continue
		}
		if l < 4 {
			continue
		}
		id := binary.LittleEndian.Uint32(data)
		segment := &Segment{
			id:   id,
			data: data[4:l],
			addr: addr,
		}
		//get session
		pSession := r.getSession(segment)
		if pSession == nil {
			continue
		}
		//check
		if pSession.checkSement(segment) == false {
			continue
		}
		r.appendSegment(segment)
		r.reply(segment)
	}
}

func (r *Receiver) getSession(segment *Segment) *session {

	if r.pSession == nil {
		r.pSession = NewSession(*segment.addr)
		return r.pSession
	}
	if r.pSession.raddr.String() != segment.addr.String() {
		return nil
	}
	return r.pSession
}

//>0掉包
var debugIdx = 0
var debugMissing = []int{0, 1}

//reply first missing segment & max segment id
func (r *Receiver) reply(segment *Segment) {
	pSession := r.pSession
	if pSession == nil {
		Log("reply session nil")
		return
	}
	data := r.getReplyData(pSession.nextId, pSession.maxId)
	Log("s.sendReply nextId:%v maxId:%v", pSession.nextId, pSession.maxId)
	if pSession.raddr.String() != r.pConn.RemoteAddr().String() {
		Log("pSession.raddr.String()!=r.pConn.RemoteAddr() return")
		return
	}
	_, err := r.pConn.Write(data)
	if CheckErr(err) {
		panic(err)
	}
}

func (r *Receiver) getReplyData(nextId uint32, maxId uint32) []byte {
	dataNext := make([]byte, 4)
	dataMax := make([]byte, 4)
	binary.LittleEndian.PutUint32(dataNext, nextId)
	binary.LittleEndian.PutUint32(dataMax, maxId)
	return append(dataNext, dataMax...)
}

//sort append by segment.id  ,append continuous data 2 recv_buf
func (r *Receiver) appendSegment(segment *Segment) {
	r.log("append segment:%v", segment)
	pSession := r.pSession
	pSession.appendAndSort(segment)
	r.log("recv data :%v", pSession.recvBuff)
	r.log("sort:%v", pSession.list)
}

func CheckErr(err error) bool {
	if err != nil {
		Log("err:%v", err.Error())
		return true
	}
	return false
}

func (r *Receiver) Read(p []byte) (l int, err error) {
	return r.pSession.Read(p)
}

File: Receiver.cs
using UnityEngine;
using System.Collections;
using System.Net.Sockets;
using System.Net;
using System.Threading;
using System;
using System.Collections.Generic;


public class Receiver  {

    IPEndPoint raddr;
    Socket socket;
    Session session;
    Thread thRecv = null;
    int timeout = 5000;

    public int DataCount
    {
        get
        {
            if (session == null)
            {
                return 0;
            }
            return session.recvBuff.Count;
        }
    }

    public void Dial(string addr,int port)
    {
        raddr = new IPEndPoint(IPAddress.Parse(addr), port);
        socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
        session = new Session(raddr);
        thRecv = new Thread(Recv);
        thRecv.Start();
    }

    public void Close() {
        try
        {
            thRecv.Abort();
        }
        catch(Exception e) {
            Log.Info("{0}", e);
        }
        finally
        {
            Log.Info("receiver close");
        }
    }
    
    void Recv()
    {
        try
        {
            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReceiveTimeout, timeout);
            var firstData = getReplyData(0, 0);
            socket.SendTo(firstData, raddr);
            for (;;)
            {
                var data = new byte[1024];
                EndPoint addr = this.raddr;
                var n = socket.ReceiveFrom(data, ref addr);
                var id = bytes2uint(data);
                var newData = new byte[n - 4];
                for (int i = 0; i < newData.Length; i++)
                {
                    newData[i] = data[i + 4];
                }
                var segment = new Segment(id, newData, this.raddr);
                var pSession = getSession(segment);
                if (pSession == null)
                {
                    continue;
                }
                if (pSession.checkSegment(segment) == false)
                {
                    continue;
                }
                appendSegment(segment);
                reply(segment);
            }
        }catch(Exception e)
        {
            Log.Info("Recv err:{0}", e);
        }
    }
    
    byte[] getReplyData(uint nextId,uint maxId)
    {
        byte[] b = new byte[8];
        uint2bytes(nextId).CopyTo(b, 0);
        uint2bytes(maxId).CopyTo(b, 4);
        return b;
    }

    uint bytes2uint(byte[] b)
    {
        if (BitConverter.IsLittleEndian == false)
        {
            Array.Reverse(b);
        }
        return BitConverter.ToUInt32(b, 0);
    }

    byte[] uint2bytes(uint i)
    {

        var b = System.BitConverter.GetBytes(i);
        if (BitConverter.IsLittleEndian == false)
        {
            Array.Reverse(b);
        }
        return b;
    }

    Session getSession(Segment segment)
    {
        if (this.session == null)
        {
            this.session = new Session(segment.addr);
            return session;
        }
        if (this.session.checkAddr(segment) == false)
        {
            return null;
        }
        return session;
    }


    void appendSegment(Segment segment)
    {
        session.appendAndSort(segment);
    }

    void reply(Segment segment) {
        if (session == null)
        {
            Log.Info("reply session nil");
            return;
        }
        var data = getReplyData(session.nextId, session.maxId);
        Log.Info("s.sendReply nextId:{0} maxId:{1}", session.nextId, session.maxId);
        if (session.raddr.ToString() != this.raddr.ToString())
        {
            Log.Info("pSession.raddr.String()!=r.pConn.RemoteAddr() return");
            return;
        }
        try
        {
            socket.SendTo(data, raddr);
        }
        catch(Exception e)
        {
            Log.Info("sendTo err:{0}", e);
        }
    }

    public byte[] Read() {
        if (this.session == null)
        {
            return null;
        }
        if (this.session.recvBuff.Count > 0)
        {
            return this.session.Read();
        }
        return null;
    }
}


public class Session
{
    public IPEndPoint raddr;
    public List<Segment> list;
    public uint nextId;
    public uint maxId;
    public List<byte> recvBuff;
    object mLock;

    public Session(IPEndPoint raddr)
    {
        this.raddr = raddr;
        list = new List<Segment>();
        recvBuff = new List<byte>();
        mLock = new object();
    }

    public bool checkSegment(Segment segment)
    {
        if (segment.id < this.nextId)
        {
            return false;
        }
        foreach (var v in list)
        {
            if (v.id == segment.id)
            {
                return false;
            }
        }
        return true;
    }

    public void appendAndSort(Segment segment)
    {
        var insertIdx = list.Count;
        for (int idx = 0; idx < list.Count; idx++)
        {
            var v = list[idx];
            if (v.id > segment.id)
            {
                insertIdx = idx;
                break;
            }
        }
        Log.Info("insertIdx:{0}", insertIdx);
        Log.Info("before:{0}", list);
        if (insertIdx == list.Count)
        {
            list.Add(segment);
        }
        else
        {
            //check
            list.Insert(insertIdx, segment);
        }
        Log.Info("after:{0}", list);
        if (maxId < segment.id)
        {
            maxId = segment.id;
        }
        for (int i = 0; i < list.Count; i++)
        {
            if (nextId == list[i].id)
            {
                lock (mLock)
                {
                    recvBuff.AddRange(list[i].data);
                }
                list.RemoveAt(0);
                nextId++;
                i--;
            }
            else
            {
                break;
            }
        }
    }

    public byte[] Read()
    {
        byte[] data = null;
        lock (mLock)
        {
            data = recvBuff.ToArray();
            recvBuff.Clear();
        }
        return data;
    }

    public bool checkAddr(Segment other)
    {
        if (other == null) return false;
        if (this.raddr.Port == other.addr.Port && this.raddr.Address.Equals(other.addr.Address))
        {
            return true;
        }
        return false;
    }

}

public class Segment
{
    public uint id;
    public byte[] data;
    public IPEndPoint addr;

    public Segment(uint id,byte[] data,IPEndPoint addr) {
        this.id = id;
        this.data = data;
        this.addr = addr;
    }

    public override string ToString()
    {
        return addr.ToString();
    }
}

public class Log
{
    public static void Info(string format,params object[] args)
    {
        return;
        Debug.Log(string.Format(format, args));
        UnityEngine.MonoBehaviour.print(string.Format(format, args));
    }
}